File: leetcode/0130-surrounded-regions/solution.cpp
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        int n = board.size();
        int m = board[0].size();
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        for(int i=0;i<n;i++) {  
            for(int j=0;j<m;j++) {
                if(vis[i][j] || board[i][j] == 'X') {
                    continue;
                }
                bool surrounded = true;
                vector<int> buffer;
                vector<int> q = {i,j};
                while(!q.empty()) {
                    int y = q.back();q.pop_back();
                    int x = q.back();q.pop_back();
                    if(x < 0 || x >= n || y < 0 || y >= m) {
                        surrounded = false;
                        continue;
                    }
                    if(vis[x][y] || board[x][y] == 'X') {
                        continue;
                    }
                    vis[x][y] = true;
                    buffer.push_back(x);
                    buffer.push_back(y);
                    q.push_back(x+1);q.push_back(y);
                    q.push_back(x-1);q.push_back(y);
                    q.push_back(x);q.push_back(y+1);
                    q.push_back(x);q.push_back(y-1);
                }
                while(surrounded && !buffer.empty()) {
                    int y = buffer.back();buffer.pop_back();
                    int x = buffer.back();buffer.pop_back();
                    board[x][y] = 'X';
                }
            }
        }
    }
};

File: leetcode/0130-surrounded-regions/solution.rs
impl Solution {
    pub fn solve(board: &mut Vec<Vec<char>>) {
        let n = board.len();
        let m = board[0].len();
        let mut vis = vec![vec![false;m];n];
        for i in 0..n {
            for j in 0..m {
                if board[i][j] == 'X' {
                    continue;
                }
                let mut surrounded = true;
                let mut q = vec![(i,j)];
                let mut buffer = Vec::new();
                while let Some((i,j)) = q.pop() {
                    if vis[i][j] || board[i][j] == 'X' {
                        continue;
                    }
                    vis[i][j] = true;
                    if i == 0 || j == 0 || i == n-1 || j == m-1 {
                        surrounded = false;
                    }
                    buffer.push((i,j));
                    if i > 0 {
                        q.push((i-1, j));
                    }
                    if i < n-1 {
                        q.push((i+1, j));
                    }
                    if j > 0 {
                        q.push((i, j-1));
                    }
                    if j < m-1 {
                        q.push((i, j+1));
                    }
                }
                if surrounded {
                    while let Some((i,j)) = buffer.pop() {
                        board[i][j] = 'X';
                    }
                }
            }
        }
    }
}
